config/config.go

```diff
@@ -290,60 +318,120 @@ func formatMetadataError(err error) error {
+		// ".google.internal", and this IP address is documented as
+		// being stable anyway.
+		host = metadataIP
+	}
+	computeMetadataURL := "http://" + host + "/computeMetadata/v1/" + suffix
+	req, err := http.NewRequest("GET", computeMetadataURL, nil)
+	if err != nil {
+		return "", "", err
+	}
+	req.Header.Add("Metadata-Flavor", "Google")
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return "", "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusNotFound {
+		return "", "", err
+	}
+	all, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", "", err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return "", "", fmt.Errorf("recevied status code %q for request %s", resp.Status, req.URL.String())
+	}
+
+	return string(all), resp.Header.Get("Etag"), nil
+}
+
+// InitConfig initializes AgentConfig and resets the lastEtag to "0".
+// InitConfig is only used during the application startup to
+// initialize the config once.
+// This is a stateless API to fetch current config from metadata server.
+func InitConfig(ctx context.Context) error {
+	if lEtag.get() != "0" {
+		return fmt.Errorf("InitConfig must not be called with non-zero lastEtag. call WatchConfg")
+	}
+	err := WatchConfig(ctx, nil)
+	if err != nil {
+		return err
+	}
+	lEtag.set("0")
+	return nil
+}
+
+// WatchConfig looks for changes in metadata keys. Upon receiving successful response,
+// it create a new agent config and executes the callback fn.
+func WatchConfig(ctx context.Context, fn func()) error {
 	var md string
 	var webError error
+	eTag := lEtag.get()
 	webErrorCount := 0
-	ticker := time.NewTicker(5 * time.Second)
 	for {
-		md, webError = metadata.Get("?recursive=true&alt=json")
-		if webError == nil {
+		md, eTag, webError = getMetadata(fmt.Sprintf("?recursive=true&alt=json&wait_for_change=true&last_etag=%s&timeout_sec=%d", lEtag.get(), osConfigMetadataPollTimeout))
+		if webError == nil && eTag != lEtag.get() {
 			break
 		}
+
 		// Try up to 3 times to wait for slow network initialization, after
 		// that resort to using defaults and returning the error.
-		if webErrorCount == 2 {
-			webError = formatMetadataError(webError)
-			break
+		if webError != nil {
+			if webErrorCount == 2 {
+				webError = formatMetadataError(webError)
+				break
+			}
+			webErrorCount++
 		}
-		webErrorCount++
+
 		select {
-		case <-ticker.C:
-			continue
 		case <-ctx.Done():
 			return nil
+		default:
+			continue
 		}
 	}
 
-	var metadata metadataJSON
-	if err := json.Unmarshal([]byte(md), &metadata); err != nil {
+	logger.Debugf("Metadata changed.")
+	lEtag.set(eTag)
+
+	var metadataConfig metadataJSON
+	if err := json.Unmarshal([]byte(md), &metadataConfig); err != nil {
 		return err
 	}
 
-	new := createConfigFromMetadata(metadata)
+	newAgentConfig := createConfigFromMetadata(metadataConfig)
 	agentConfigMx.Lock()
-	agentConfig = new
+	agentConfig = newAgentConfig
 	agentConfigMx.Unlock()
 
+	logFeatures()
```

Comment:
Don't do that here, the caller should probably handle this.
We should also only log if there is a change, so for now just skip this.

Reply:
i had the same feelings that we should only log if there is change in osconfig features, since diff based actions might be a little complicated. since this is an explicit ask from customer, i think its better to start with logging current state of the world, and moving forward we can change this to diff based.

also, since the feature list belonged to config library, i thought keeping the logging in config library would be best.

Reply:
ok.. i think it makes sense to move the call to called of watchconfig. sorry for misunderstanding.